File: src/services/api.service.ts
import fetch from 'node-fetch';
import { URLSearchParams } from 'url';
import { getAuthUrl, getFetchLocalesUrl, getCreateKeyUrl, getUpdateKeyUrl, getSearchKeysUrl, getProjectsUrl, getUserInfoUrl } from '../constants';
import { AuthRequest, AuthResponse, AuthErrorResponse, FetchLocalesResponse, CreateKeyRequest, CreateKeyResponse, CreateKeyErrorResponse, UpdateKeyResponse, SearchKeysResponse, ProjectsResponse } from './api.service.types';
import { StorageService } from './storage.service';

export class ApiService {
    private readonly CLIENT_ID = 'auth-client';
    private readonly CLIENT_SECRET = '77X6lydjx3beOw3rcgQ6fwMK6t2tatFo';
    private readonly GRANT_TYPE_PASSWORD = 'password';
    private readonly GRANT_TYPE_REFRESH = 'refresh_token';
    private readonly SCOPE = 'openid';
    private storageService?: StorageService;
    private isRefreshing = false;
    private refreshPromise: Promise<AuthResponse | null> | null = null;

    constructor(storageService?: StorageService) {
        this.storageService = storageService;
    }

    async authenticate(username: string, password: string): Promise<AuthResponse> {
        const formBody = this.buildAuthBody(username, password);
        
        const res = await fetch(await getAuthUrl(), {
            headers: {
                "endpoint": "KC_AUTH",
                "content-type": "application/x-www-form-urlencoded;charset=UTF-8",
            },
            body: formBody.toString(),
            method: "POST"
        });

        if (!res.ok) {
            try {
                const errorData: AuthErrorResponse = await res.json();
                const errorMessage = errorData.error_description || errorData.error || 'Ошибка авторизации';
                throw new Error(errorMessage);
            } catch (parseError) {
                if (parseError instanceof Error && parseError.message !== 'Unexpected end of JSON input') {
                    throw parseError;
                }
                throw new Error(`Ошибка авторизации: ${res.status} ${res.statusText}`);
            }
        }

        const data: AuthResponse = await res.json();
        return data;
    }

    async refreshToken(refreshToken: string): Promise<AuthResponse> {
        const params = new URLSearchParams();
        params.append('grant_type', this.GRANT_TYPE_REFRESH);
        params.append('refresh_token', refreshToken);
        params.append('client_id', this.CLIENT_ID);
        params.append('client_secret', this.CLIENT_SECRET);

        const res = await fetch(await getAuthUrl(), {
            headers: {
                "endpoint": "KC_AUTH",
                "content-type": "application/x-www-form-urlencoded;charset=UTF-8",
            },
            body: params.toString(),
            method: "POST"
        });

        if (!res.ok) {
            try {
                const errorData: AuthErrorResponse = await res.json();
                const errorMessage = errorData.error_description || errorData.error || 'Ошибка обновления токена';
                throw new Error(errorMessage);
            } catch (parseError) {
                if (parseError instanceof Error && parseError.message !== 'Unexpected end of JSON input') {
                    throw parseError;
                }
                throw new Error(`Ошибка обновления токена: ${res.status} ${res.statusText}`);
            }
        }

        const data: AuthResponse = await res.json();
        return data;
    }

    async getUserInfo(token: string): Promise<any> {
        const url = await getUserInfoUrl();
        const res = await fetch(url, {
            headers: {
                'authorization': `Bearer ${token}`,
            },
            method: 'GET'
        });

        if (!res.ok) {
            throw new Error(`Failed to get user info: ${res.status}`);
        }

        return await res.json();
    }

    private async ensureValidToken(): Promise<string> {
        if (!this.storageService) {
            throw new Error('StorageService не настроен');
        }

        const accessToken = await this.storageService.getAccessToken();
        if (!accessToken) {
            throw new Error('Токен не найден');
        }

        return accessToken;
    }

    private async refreshTokenIfNeeded(): Promise<string> {
        if (!this.storageService) {
            throw new Error('StorageService не настроен');
        }

        if (this.isRefreshing && this.refreshPromise) {
            const result = await this.refreshPromise;
            if (result) {
                return result.access_token;
            }
            throw new Error('Не удалось обновить токен');
        }

        this.isRefreshing = true;
        this.refreshPromise = (async () => {
            try {
                const refreshToken = await this.storageService!.getRefreshToken();
                if (!refreshToken) {
                    await this.storageService!.deleteTokens();
                    throw new Error('Refresh token не найден');
                }

                const authResponse = await this.refreshToken(refreshToken);
                await this.storageService!.saveTokens(authResponse);
                return authResponse;
            } catch (error) {
                await this.storageService!.deleteTokens();
                throw error;
            } finally {
                this.isRefreshing = false;
                this.refreshPromise = null;
            }
        })();

        const result = await this.refreshPromise;
        if (result) {
            return result.access_token;
        }
        throw new Error('Не удалось обновить токен');
    }

    private async makeRequestWithTokenRefresh<T>(
        requestFn: (token: string) => Promise<T>
    ): Promise<T> {
        let token = await this.ensureValidToken();

        try {
            return await requestFn(token);
        } catch (error: any) {
            if (error.status === 401 || (error.message && error.message.includes('401'))) {
                token = await this.refreshTokenIfNeeded();
                return await requestFn(token);
            }
            throw error;
        }
    }

    async fetchProjects(token?: string): Promise<ProjectsResponse> {
        if (this.storageService && !token) {
            return this.makeRequestWithTokenRefresh((t) => this.fetchProjects(t));
        }

        const actualToken = token || await this.ensureValidToken();
        const url = await getProjectsUrl();
        const res = await fetch(url, {
            headers: {
                'authorization': `Bearer ${actualToken}`,
            },
            method: 'GET'
        });

        if (!res.ok) {
            const error = new Error(`Failed to fetch projects: ${res.status}`);
            (error as any).status = res.status;
            throw error;
        }

        const data: ProjectsResponse = await res.json();
        return data;
    }

    async fetchLocales(token: string | undefined, locale: string, projectKey: string): Promise<Record<string, string>> {
        if (this.storageService && !token) {
            return this.makeRequestWithTokenRefresh((t) => this.fetchLocales(t, locale, projectKey));
        }

        const actualToken = token || await this.ensureValidToken();
        const url = await getFetchLocalesUrl(locale, projectKey);
        const res = await fetch(url, {
            headers: {
                Authorization: `Bearer ${actualToken}`
            }
        });

        if (!res.ok) {
            const error = new Error(`Failed to fetch locales for ${locale}: ${res.status}`);
            (error as any).status = res.status;
            throw error;
        }

        const data: FetchLocalesResponse = await res.json();
        return data.data;
    }

    async createKey(token: string | undefined, request: CreateKeyRequest, projectKey: string): Promise<CreateKeyResponse> {
        if (this.storageService && !token) {
            return this.makeRequestWithTokenRefresh((t) => this.createKey(t, request, projectKey));
        }

        const actualToken = token || await this.ensureValidToken();
        const url = await getCreateKeyUrl(projectKey);
        const res = await fetch(url, {
            method: 'POST',
            headers: {
                'authorization': `Bearer ${actualToken}`,
                'content-type': 'application/json',
            },
            body: JSON.stringify(request)
        });

        console.log(url)
        console.log(JSON.stringify(request))

        const data = await res.json();

        if (!res.ok || 'message' in data) {
            const error = data as CreateKeyErrorResponse;
            if (error.message.includes('duplicate key')) {
                throw new Error('Ключ уже существует');
            }
            const err = new Error(error.message || `Failed to create key: ${res.status}`);
            (err as any).status = res.status;
            throw err;
        }

        return data as CreateKeyResponse;
    }

    async updateKey(token: string | undefined, request: CreateKeyRequest, projectKey: string): Promise<UpdateKeyResponse> {
        if (this.storageService && !token) {
            return this.makeRequestWithTokenRefresh((t) => this.updateKey(t, request, projectKey));
        }

        const actualToken = token || await this.ensureValidToken();
        const url = await getUpdateKeyUrl(projectKey);
        const res = await fetch(url, {
            method: 'POST',
            headers: {
                'authorization': `Bearer ${actualToken}`,
                'content-type': 'application/json',
            },
            body: JSON.stringify(request)
        });

        const text = await res.text();
        
        if (!text || text.trim().length === 0) {
            if (res.ok) {
                return {
                    data: {
                        key: request.key,
                        translations: request.translations
                    }
                };
            }
            const err = new Error(`Failed to update key: ${res.status} (empty response)`);
            (err as any).status = res.status;
            throw err;
        }

        let data: any;
        try {
            data = JSON.parse(text);
        } catch (e) {
            throw new Error(`Invalid JSON response: ${text.substring(0, 100)}`);
        }

        if (!res.ok || 'message' in data) {
            const error = data as CreateKeyErrorResponse;
            const err = new Error(error.message || `Failed to update key: ${res.status}`);
            (err as any).status = res.status;
            throw err;
        }

        return data as UpdateKeyResponse;
    }

    async searchKeys(token: string | undefined, search: string, projectKey: string, pageNumber: number = 1, pageSize: number = 10): Promise<SearchKeysResponse> {
        if (this.storageService && !token) {
            return this.makeRequestWithTokenRefresh((t) => this.searchKeys(t, search, projectKey, pageNumber, pageSize));
        }

        const actualToken = token || await this.ensureValidToken();
        const params = new URLSearchParams();
        params.append('pageNumber', pageNumber.toString());
        params.append('pageSize', pageSize.toString());
        if (search && search.trim()) {
            params.append('search', search.trim());
        }

        const url = `${await getSearchKeysUrl(projectKey)}?${params.toString()}`;
        const res = await fetch(url, {
            method: 'GET',
            headers: {
                'authorization': `Bearer ${actualToken}`,
            }
        });

        if (!res.ok) {
            const error = new Error(`Failed to search keys: ${res.status}`);
            (error as any).status = res.status;
            throw error;
        }

        const data = await res.json();
        return data as SearchKeysResponse;
    }

    private buildAuthBody(username: string, password: string): URLSearchParams {
        const req: AuthRequest = {
            username,
            password,
            client_id: this.CLIENT_ID,
            client_secret: this.CLIENT_SECRET,
            grant_type: this.GRANT_TYPE_PASSWORD,
            scope: this.SCOPE
        };

        const params = new URLSearchParams();
        params.append('username', req.username);
        params.append('password', req.password);
        params.append('client_id', req.client_id);
        params.append('client_secret', req.client_secret);
        params.append('grant_type', req.grant_type);
        params.append('scope', req.scope);
        
        return params;
    }
}

File: src/ide/webstorm/src/main/java/com/point/i18n/webstorm/services/ApiService.java
package com.point.i18n.webstorm.services;

import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.IOException;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;

/**
 * Service for API interactions.
 * Simplified Java implementation of the TypeScript ApiService.
 */
public class ApiService {
    private static final String CLIENT_ID = "auth-client";
    private static final String CLIENT_SECRET = "77X6lydjx3beOw3rcgQ6fwMK6t2tatFo";
    private static final String GRANT_TYPE_PASSWORD = "password";
    private static final String SCOPE = "openid";
    
    private final StorageService storageService;
    private final ConfigService configService;
    private final ObjectMapper objectMapper;
    
    {
        // Настраиваем ObjectMapper для игнорирования неизвестных полей
        objectMapper = new ObjectMapper();
        objectMapper.configure(com.fasterxml.jackson.databind.DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
    }
    
    public ApiService(StorageService storageService) {
        this.storageService = storageService;
        this.configService = new ConfigService();
    }
    
    @com.fasterxml.jackson.annotation.JsonIgnoreProperties(ignoreUnknown = true)
    public static class AuthResponse {
        public String access_token;
        public String refresh_token;
        public int refresh_expires_in;
        public Integer expires_in; // Может быть в ответе, но не используется
    }
    
    public AuthResponse authenticate(String username, String password) throws IOException {
        String apiBaseUrl = configService.getApiBaseUrl();
        String authUrl = apiBaseUrl + "/api/v1/proxy/realms/auth/protocol/openid-connect/token";
        
        // Формируем тело запроса
        String formBody = String.format(
            "grant_type=%s&username=%s&password=%s&client_id=%s&client_secret=%s&scope=%s",
            GRANT_TYPE_PASSWORD,
            java.net.URLEncoder.encode(username, StandardCharsets.UTF_8),
            java.net.URLEncoder.encode(password, StandardCharsets.UTF_8),
            CLIENT_ID,
            CLIENT_SECRET,
            SCOPE
        );
        
        // Выполняем запрос
        URL url = new URL(authUrl);
        HttpURLConnection connection = (HttpURLConnection) url.openConnection();
        connection.setRequestMethod("POST");
        connection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded;charset=UTF-8");
        connection.setRequestProperty("endpoint", "KC_AUTH");
        connection.setDoOutput(true);
        
        try (java.io.OutputStream os = connection.getOutputStream()) {
            byte[] input = formBody.getBytes(StandardCharsets.UTF_8);
            os.write(input, 0, input.length);
        }
        
        // Читаем ответ
        int responseCode = connection.getResponseCode();
        if (responseCode != HttpURLConnection.HTTP_OK) {
            String errorMessage = "Authentication failed: " + responseCode;
            try (java.io.BufferedReader reader = new java.io.BufferedReader(
                    new java.io.InputStreamReader(connection.getErrorStream(), StandardCharsets.UTF_8))) {
                StringBuilder response = new StringBuilder();
                String line;
                while ((line = reader.readLine()) != null) {
                    response.append(line);
                }
                if (response.length() > 0) {
                    errorMessage = response.toString();
                }
            }
            throw new IOException(errorMessage);
        }
        
        try (java.io.BufferedReader reader = new java.io.BufferedReader(
                new java.io.InputStreamReader(connection.getInputStream(), StandardCharsets.UTF_8))) {
            StringBuilder response = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) {
                response.append(line);
            }
            
            return objectMapper.readValue(response.toString(), AuthResponse.class);
        }
    }
    
    // Response classes
    @com.fasterxml.jackson.annotation.JsonIgnoreProperties(ignoreUnknown = true)
    public static class ProjectsResponse {
        public java.util.List<Project> data;
        public int totalCount;
        public int totalPages;
    }
    
    @com.fasterxml.jackson.annotation.JsonIgnoreProperties(ignoreUnknown = true)
    public static class Project {
        public String key;
        public String name;
    }
    
    @com.fasterxml.jackson.annotation.JsonIgnoreProperties(ignoreUnknown = true)
    public static class FetchLocalesResponse {
        public Map<String, String> data;
    }
    
    @com.fasterxml.jackson.annotation.JsonIgnoreProperties(ignoreUnknown = true)
    public static class CreateKeyRequest {
        public String key;
        public Translations translations;
    }
    
    @com.fasterxml.jackson.annotation.JsonIgnoreProperties(ignoreUnknown = true)
    public static class Translations {
        public String ru;
        public String en;
        public String uz;
    }
    
    @com.fasterxml.jackson.annotation.JsonIgnoreProperties(ignoreUnknown = true)
    public static class CreateKeyResponse {
        public KeyData data;
    }
    
    @com.fasterxml.jackson.annotation.JsonIgnoreProperties(ignoreUnknown = true)
    public static class KeyData {
        public String key;
        public Translations translations;
    }
    
    @com.fasterxml.jackson.annotation.JsonIgnoreProperties(ignoreUnknown = true)
    public static class SearchKeysResponse {
        public java.util.List<KeyData> data;
        public int totalCount;
        public int totalPages;
    }
    
    // Helper method to get token
    private String getToken() throws IOException {
        String token = storageService.getAccessToken();
        if (token == null || token.isEmpty()) {
            throw new IOException("Token not found");
        }
        return token;
    }
    
    // Helper method to make GET request
    private String makeGetRequest(String url, String token) throws IOException {
        URL urlObj = new URL(url);
        HttpURLConnection connection = (HttpURLConnection) urlObj.openConnection();
        connection.setRequestMethod("GET");
        connection.setRequestProperty("Authorization", "Bearer " + token);
        
        int responseCode = connection.getResponseCode();
        if (responseCode != HttpURLConnection.HTTP_OK) {
            throw new IOException("Request failed: " + responseCode);
        }
        
        try (java.io.BufferedReader reader = new java.io.BufferedReader(
                new java.io.InputStreamReader(connection.getInputStream(), StandardCharsets.UTF_8))) {
            StringBuilder response = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) {
                response.append(line);
            }
            return response.toString();
        }
    }
    
    // Helper method to make POST request
    private String makePostRequest(String url, String token, String jsonBody) throws IOException {
        URL urlObj = new URL(url);
        HttpURLConnection connection = (HttpURLConnection) urlObj.openConnection();
        connection.setRequestMethod("POST");
        connection.setRequestProperty("Authorization", "Bearer " + token);
        connection.setRequestProperty("Content-Type", "application/json");
        connection.setDoOutput(true);
        
        try (java.io.OutputStream os = connection.getOutputStream()) {
            byte[] input = jsonBody.getBytes(StandardCharsets.UTF_8);
            os.write(input, 0, input.length);
        }
        
        int responseCode = connection.getResponseCode();
        if (responseCode != HttpURLConnection.HTTP_OK && responseCode != HttpURLConnection.HTTP_CREATED) {
            String errorMessage = "Request failed: " + responseCode;
            try (java.io.BufferedReader reader = new java.io.BufferedReader(
                    new java.io.InputStreamReader(connection.getErrorStream(), StandardCharsets.UTF_8))) {
                StringBuilder response = new StringBuilder();
                String line;
                while ((line = reader.readLine()) != null) {
                    response.append(line);
                }
                if (response.length() > 0) {
                    errorMessage = response.toString();
                }
            }
            throw new IOException(errorMessage);
        }
        
        try (java.io.BufferedReader reader = new java.io.BufferedReader(
                new java.io.InputStreamReader(connection.getInputStream(), StandardCharsets.UTF_8))) {
            StringBuilder response = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) {
                response.append(line);
            }
            return response.length() > 0 ? response.toString() : "{}";
        }
    }
    
    public ProjectsResponse fetchProjects() throws IOException {
        String token = getToken();
        String apiBaseUrl = configService.getApiBaseUrl();
        String url = apiBaseUrl + "/api/v1/proxy/localization/api/localization-project?pageSize=100";
        String response = makeGetRequest(url, token);
        return objectMapper.readValue(response, ProjectsResponse.class);
    }
    
    public Map<String, String> fetchLocales(String locale, String projectKey) throws IOException {
        String token = getToken();
        String localizationApiBaseUrl = configService.getLocalizationApiBaseUrl();
        String url = localizationApiBaseUrl + "/api/v1/localization/" + projectKey + "/language/" + locale;
        String response = makeGetRequest(url, token);
        FetchLocalesResponse localesResponse = objectMapper.readValue(response, FetchLocalesResponse.class);
        return localesResponse.data != null ? localesResponse.data : new HashMap<>();
    }
    
    public CreateKeyResponse createKey(CreateKeyRequest request, String projectKey) throws IOException {
        String token = getToken();
        String apiBaseUrl = configService.getApiBaseUrl();
        String url = apiBaseUrl + "/api/v1/proxy/localization/api/localization/" + projectKey + "/new";
        String jsonBody = objectMapper.writeValueAsString(request);
        String response = makePostRequest(url, token, jsonBody);
        return objectMapper.readValue(response, CreateKeyResponse.class);
    }
    
    public CreateKeyResponse updateKey(CreateKeyRequest request, String projectKey) throws IOException {
        String token = getToken();
        String apiBaseUrl = configService.getApiBaseUrl();
        String url = apiBaseUrl + "/api/v1/proxy/localization/api/localization/" + projectKey;
        String jsonBody = objectMapper.writeValueAsString(request);
        String responseText = makePostRequest(url, token, jsonBody);
        
        // Обрабатываем пустой ответ или null data
        if (responseText == null || responseText.trim().isEmpty() || responseText.trim().equals("{}")) {
            // Если ответ пустой, но запрос успешен, возвращаем данные из request
            CreateKeyResponse response = new CreateKeyResponse();
            response.data = new KeyData();
            response.data.key = request.key;
            response.data.translations = request.translations;
            return response;
        }
        
        CreateKeyResponse response = objectMapper.readValue(responseText, CreateKeyResponse.class);
        
        // Если data == null, создаем из request
        if (response.data == null) {
            response.data = new KeyData();
            response.data.key = request.key;
            response.data.translations = request.translations;
        }
        
        return response;
    }
    
    public SearchKeysResponse searchKeys(String search, String projectKey, int pageNumber, int pageSize) throws IOException {
        String token = getToken();
        String apiBaseUrl = configService.getApiBaseUrl();
        StringBuilder urlBuilder = new StringBuilder(apiBaseUrl + "/api/v1/proxy/localization/api/localization/" + projectKey);
        urlBuilder.append("?pageNumber=").append(pageNumber);
        urlBuilder.append("&pageSize=").append(pageSize);
        if (search != null && !search.trim().isEmpty()) {
            urlBuilder.append("&search=").append(java.net.URLEncoder.encode(search.trim(), StandardCharsets.UTF_8));
        }
        String url = urlBuilder.toString();
        String response = makeGetRequest(url, token);
        return objectMapper.readValue(response, SearchKeysResponse.class);
    }
}

